session/txn.go

```diff
@@ -386,20 +288,19 @@ func (st *TxnState) cleanup() {
 
 // KeysNeedToLock returns the keys need to be locked.
 func (st *TxnState) KeysNeedToLock() ([]kv.Key, error) {
-	if st.stmtBufLen() == 0 {
+	if st.stagingHandle == kv.InvalidStagingHandle {
 		return nil, nil
 	}
-	keys := make([]kv.Key, 0, st.stmtBufLen())
-	if err := kv.WalkMemBuffer(st.stmtBuf(), func(k kv.Key, v []byte) error {
+	keys := make([]kv.Key, 0, st.countHint())
+	buf := st.Transaction.GetMemBuffer()
+	buf.InspectStage(st.stagingHandle, func(k kv.Key, flags kv.KeyFlags, v []byte) {
 		if !keyNeedToLock(k, v) {
-			return nil
+			return
 		}
 		// If the key is already locked, it will be deduplicated in LockKeys method later.
 		keys = append(keys, k)
-		return nil
-	}); err != nil {
-		return nil, err
-	}
+	})
+	// LockKeys will sort and deduplicate keys.
```

Comment:
why add this comment?

Reply:
Keys can only be returned from InspectStage according to the order of writing.

Reply:
`LockKeys` already handled this in

https://github.com/pingcap/tidb/blob/6cf50d3f898cadf3e74adaa6c12ab0e671bd9104/store/tikv/txn.go#L437-L448

Reply:
Yes, it already has a comment,
https://github.com/pingcap/tidb/pull/18657/files#diff-417f1f41cb3111fe7a10aa64ea77d1efR300